src/translator/vocabs.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+namespace marian {
+namespace bergamot {
+
+class VocabsGenerator {
+public:
+  VocabsGenerator(Ptr<Options> options): options_(options){}
+
+  // load from buffer
+  void load(std::vector<std::shared_ptr<AlignedMemory>>&& vocabMemories) {
+    // At least two vocabs: src and trg
+    ABORT_IF(vocabMemories.size() < 2, "Insufficient number of vocabularies.");
+    vocabs_.resize(vocabMemories.size());
+    // hashMap is introduced to avoid double loading the same vocab
+    // loading vocabs (either from buffers or files) is the biggest bottleneck of the speed
+    // uintptr_t holds unique keys (address) for share_ptr<AlignedMemory>
+    std::unordered_map<uintptr_t, Ptr<Vocab>> vmap;
+    for (size_t i = 0; i < vocabs_.size(); i++) {
+      auto m = vmap.emplace(std::make_pair(reinterpret_cast<uintptr_t>(vocabMemories[i].get()), Ptr<Vocab>()));
+      if (m.second) { // new: load the vocab
+        m.first->second = New<Vocab>(options_, i);
+        m.first->second->loadFromSerialized(absl::string_view(vocabMemories[i]->begin(), vocabMemories[i]->size()));
+      }
+      vocabs_[i] = m.first->second;
+    }
+  }
+
+  // load from file
+  void load(const std::vector<std::string>& vocabPaths){
+    // with the current setup, we need at least two vocabs: src and trg
+    ABORT_IF(vocabPaths.size() < 2, "Insufficient number of vocabularies.");
+    vocabs_.resize(vocabPaths.size());
+    std::unordered_map<std::string, Ptr<Vocab>> vmap;
+    for (size_t i = 0; i < vocabs_.size(); ++i) {
+      auto m = vmap.emplace(std::make_pair(vocabPaths[i], Ptr<Vocab>()));
+      if (m.second) { // new: load the vocab
+        m.first->second = New<Vocab>(options_, i);
+        m.first->second->load(vocabPaths[i]);
+      }
+      vocabs_[i] = m.first->second;
+    }
+  }
+
+  std::vector<Ptr<Vocab const>> getVocabs(){
+    return vocabs_;
+  }
+
+private:
+  std::vector<Ptr<Vocab const>> vocabs_;
+  Ptr<Options> options_;
+};
```

Comment:
This class can have two methods - `source()` and `target()` which provides me source and target vocabulary. At places, I'm using the following syntax. I'd rather this knowledge be contained within this class.

```cpp
srcVocab = vocabs.front()
tgtVocab = vocabs.back()
```

This could be contained here.

Reply:
This is bad practise in general, as you could have multiple source vocabularies. We don't technically support it, as we assume only 2 vocabularies, but this is not necessarily true!

Reply:
I understand what you're saying. But interspersing use of `vocabs.front()`, `vocabs[0]`, `vocabs.back()`, `vocabs[1]` in code is bad practice. If I want source, I should be calling `vocabs.source()`. If I want `target`, I should be calling `vocabs.target()`. 

If I have multiple sources or targets, I attach size_t idx or maybe another method to specify the i-th source-vocabulary or the i-th target vocabulary. Giving me a list of Vocabs having to go look 3 files down to figure out which one's which is bad practice. The `Vocabs` class can vary it's internals while the external API stays (maybe) stable and the rest of the wiring in marian::bergamot works irrespective of how if multiple source-vocabs or target-vocabs come in.

Reply:
Vocas source in this case should return a vector and target could return only one vocab. Are you ok with this?

Reply:
> Vocabs source in this case should return a vector and target could return only one vocab. Are you ok with this?

Ideally not. For text-proc, at places I'm only using the first-vocab. And somewhere in PR review, it's mentioned that all the vocabularies Mozilla encounters will have only two vocabs (source and target, which can be same). I'm not saying the following should be implemented, but to communicate with a loose sketch:

```
Vocabs::source() -> Only one source vocab
Vocabs::multiSource(size_t idx) -> idx-th source vocab
Vocabs::multiSources() -> All source vocabs as a vector (This is what you're suggesting)
Vocabs::target() -> Only one target vocab
Vocabs::multiTargets() -> All target-vocabs as a vector
```

The above can still work with `vector<Ptr<Vocab const>>` under the hood, except the documentations/assumptions baked into code rather than convention like first N being source and last 1 being target.

Far as I understand, the multi-source is a training only setup I'm unsure how it can be used here (bergamot/marian-server is pure inference).

Reply:
Multi source can happen when you decode with ensembles. Every model in an ensemble could have a different source vocab, provided they all have the same target vocab. You may only be using the first vocab in text processing, but that means we just don't support ensembles at all. 

You can only ever have one Target  vocabulary during decoding and multiple Source vocabularies (someone correct me if I am wrong) In general MT practitioners are aware of that.

if we say that we never ever support ensembles, then you can make the front() and back() assumptions.

Reply:
Let's keep `vector<Ptr<Vocab const>>`. What @jerinphilip suggested is good for the readability, but it might be too early to settle. We should allow multiple Source vocabularies for now not to break what Marian supports. (I will add documentation for `vector<Ptr<Vocab const>>`)

Reply:
> Vocas source in this case should return a vector

After an offline conversation, I'm ok with this approach, having made aware of https://github.com/marian-nmt/marian-dev/blob/fe74576dc3e957f97b4842951fb6ee7ec52fa324/src/translator/translator.h#L242-L243 and https://github.com/marian-nmt/marian-dev/blob/fe74576dc3e957f97b4842951fb6ee7ec52fa324/src/translator/translator.h#L303-L304

The call-site of `vocabs.front()` at TextProcessor (sentencepiece encodeWithByteRanges) will eventually need to be updated with multiple vocabs if and when we're ensembling. For now it can be `vocabs.source().front()` with a comment describing the situation.